eclair-core/src/main/scala/fr/acinq/eclair/channel/InteractiveTxBuilder.scala

```diff
@@ -0,0 +1,722 @@
+          val outputs = if (fundingParams.isInitiator) {
+            // If the initiator doesn't want to contribute, we should cancel out the dust amount artificially added previously.
+            val initiatorChangeOutputs = if (fundingParams.localAmount == 0.sat) {
+              changeOutputs.map(o => o.copy(amount = o.amount + fundingParams.dustLimit))
+            } else {
+              changeOutputs
+            }
+            // The initiator is responsible for adding the shared output.
+            TxAddOutput(fundingParams.channelId, generateSerialId(), fundingParams.fundingAmount, fundingParams.fundingPubkeyScript) +: initiatorChangeOutputs
+          } else {
+            // The protocol only requires the non-initiator to pay the fees for its inputs and outputs, discounting the
+            // common fields (shared output, version, nLockTime, etc). However, this is really hard to compute here,
+            // because we don't know the witness size of our inputs (we let bitcoind handle that). For simplicity's sake,
+            // we simply accept that we'll slightly overpay the fee (which speeds up channel confirmation).
+            changeOutputs
+          }
+          log.info("added {} inputs and {} outputs to interactive tx", inputDetails.usableInputs.length, outputs.length)
+          // We unlock the unusable inputs from previous iterations (if any) as they can be used outside of this protocol.
+          unlock(unusableInputs)
+          buildTx(FundingContributions(inputDetails.usableInputs, outputs))
+        } else {
+          // Some wallet inputs are unusable, so we must fund again to obtain usable inputs instead.
+          log.info("retrying funding as some utxos cannot be used for interactive-tx construction: {}", inputDetails.unusableInputs.map(o => s"${o.txid}:${o.index}").mkString(","))
+          val sanitizedTx = fundedTx.copy(
+            txIn = fundedTx.txIn.filter(txIn => !inputDetails.unusableInputs.contains(txIn.outPoint)),
+            // We remove the change output added by this funding iteration.
+            txOut = fundedTx.txOut.filter(txOut => txOut.publicKeyScript == fundingParams.fundingPubkeyScript),
+          )
+          fund(sanitizedTx, inputDetails.usableInputs, unusableInputs ++ inputDetails.unusableInputs)
+        }
+      case WalletFailure(t) =>
+        log.error("could not get input details: ", t)
+        // We use a generic exception and don't send the internal error to the peer.
+        replyTo ! LocalFailure(ChannelFundingError(fundingParams.channelId))
+        unlockAndStop(fundedTx.txIn.map(_.outPoint).toSet ++ unusableInputs)
+      case msg: ReceiveMessage =>
+        timers.startSingleTimer(msg, 1 second)
+        Behaviors.same
+      case Abort =>
+        timers.startSingleTimer(Abort, 1 second)
+        Behaviors.same
+    }
+  }
+
+  private def getInputDetails(txIn: TxIn, currentInputs: Seq[TxAddInput]): Future[Either[OutPoint, TxAddInput]] = {
```

Comment:
I think a `Either[TxIn, TxAddInput]` return type would be more consistent (why switch from input to output?):

```suggestion
  /**
   *
   * @param txIn
   * @param currentInputs existing valid inputs, we already know the details for those
   * @return either rejected utxos (left) or [[TxAddInput]] commands (right)
   */
  private def getInputDetails(txIn: TxIn, currentInputs: Seq[TxAddInput]): Future[Either[TxIn, TxAddInput]] = {
```

Reply:
I improved this in https://github.com/ACINQ/eclair/pull/2273/commits/83e51e46c9bf4e6617b7c7d77ce2af24da41d6de by creating an `UnusableInput` class, I think it makes it much more clear.